### ISSUE TITLE
XCFramework & Catalyst Build

### ISSUE THREAD

**Issue by jeremycochoy** [AudioKit/AudioKit#2040]
Hello,

My App should support iOS/iPad / Emulator and Catalyst support. Unfortunately, AudioKit pod's do not support mac catalyst. I didn't went into the details but I can understand that the build process and the CocoaPod system is complex and require a lot of work, so I definitively do not blame you for this.

Since I moved my libraries to the XCFramework I felt that using this setup for linking AudioKit would be the easiest way to support this 3 architectures (mostly because I know understand a little the build process). I rebuild the 4.9.5 branch with the XCFramework support. When I linked with catalyst support, I got the following error:

![image](https://user-images.githubusercontent.com/4193259/80484982-4a52e100-8961-11ea-8be7-d0da31c47311.png)

(Screenshot from a SO question here: https://stackoverflow.com/questions/61308037/compilation-error-when-porting-audiokit-ios-app-to-catalyst-macos-app/61479464#61479464 )

I am not sure weither it is an error in the build configuration of AudioKit or if it is caused by a bug in XCodebuild (This is something I will lt you figure out) but it happens that the 
SwiftInterface exported in the framework for the Catalyst Architecture is bugged.

Fortunately, it can be fixed by hand with a plain text editor. For my usage, I had to make the following change by hand:

* Remove the empty `.init` initializer in AKSampler interface:
  Go to ```@objc open class AKSampler : AKPolyphonicNode, AKComponent {``` section and remove the line `@objc override dynamic public init()`.
* In the `enableMIDI` prototypes, replace ` = midi.client,` by ` = AudioKit.midi.client,`.
* Replace `(node: AKNode? = output,` by `(node: AKNode? = AudioKit.output,` in `AKNodeRecorder` prototype.

This process will apply to the different swift interface files you are using. For example `x86_64-apple-ios-macabi.swiftinterface`.

I do not pretend that this is sufficient to fix all the problem that arrised during the export. it seams that all this can be fixed by renaming the AudioKit class to something else. Since this is a major breaking change, some peoples may prefer to edit the XCFramework and fix the swift interface by hand. I choose to let the information here so that both your team and the AudioKit user knows about this problem and the corresponding workaround. It may also be helpful for your team when you will come to officially support XCFramework.

@megastep

**Comment by megastep** [comment]
You should use our `v5-develop` branch for proper Catalyst support. We also have a beta pod you can use from our beta pod repo at https://github.com/AudioKit/Specs.

The problem is the Swift interface files need to be properly fixed and as a result we had to rename some classes in v5. But it's an easy migration.

**Comment by megastep** [comment]
FYI you are correct in your assessment and we basically renamed the `AudioKit` class to `AKManager` in v5 for this purpose.

**Comment by 0xmtn** [comment]
Hello,

I'm trying to use AudioKit with React-Native on Mac using Catalyst, but couldn't be able to find out a way around the error below: 
![image](https://user-images.githubusercontent.com/732147/81322882-429be680-909d-11ea-92f6-e9af329432d8.png)

I've tried it with:
- installing the latest stable version (v4.9.5) via pod
- building from source
- using catalyst builds from xcframework.zip
- and lastly, as you guys suggested above, pod installing v5.0.b1.
All returns the same error.

Without AudioKit, React-Native builds successfully with the warning on the screenshot above. The only difference is the error.

Could you point me in the right direction?

**Comment by megastep** [comment]
Looking at that last screenshot I think you might still be using the AKSettings class which was renamed to AKManager in the AK 5.0 beta. Maybe look for that?

**Comment by 0xmtn** [comment]
At this stage I haven't implemented AudioKit at all. This is happening at the build right after adding the framework / installing the pod.

**Comment by 0xmtn** [comment]
Hello again @megastep ,

Have you had a chance to look into this? Is this an AudioKit related problem, or is it related to smth else? 

Do you think I should open a new issue for this?

**Comment by megastep** [comment]
I haven't had a chance to take a close look at this yet, but it looks to me like a possible mismatch in Xcode versions maybe? If you use the 5.0 beta pod it shouldn't reference AKSettings. And these other errors might be coming from using a different version of Swift, so maybe using an older version of Xcode?

**Comment by 0xmtn** [comment]
I'm using Xcode 11.4 with Swift version 5.0.

**Comment by 0xmtn** [comment]
Hello @megastep ,

I got AudioKit v4.9.5 and v5.0.b1 to be successfully built with Catalyst in an Objective-C project with React-Native on top of it.

Now I'm trying to just build an example audio I/O app, copying the code from AudioKit Examples; to input sound from a headphone mic and output it from headphone speakers. Yet I get no output sound, netiher from mic nor from oscillator. 

I'm getting these errors:

```
2020-05-13 00:16:34.966261+0300 ObjCTrial2[22685:368459] [User Defaults] Couldn't write values for keys (
    ApplicationAccessibilityEnabled
) in CFPrefsPlistSource<0x600002c04b80> (Domain: com.apple.Accessibility, User: kCFPreferencesCurrentUser, ByHost: No, Container: (null), Contents Need Refresh: Yes): setting preferences outside an application's container requires user-preference-write or file-write-data sandbox access
2020-05-13 00:16:34.970072+0300 ObjCTrial2[22685:368459] [User Defaults] Couldn't write values for keys (
    FullKeyboardAccessFocusRingEnabled
) in CFPrefsPlistSource<0x600002c04b80> (Domain: com.apple.Accessibility, User: kCFPreferencesCurrentUser, ByHost: No, Container: (null), Contents Need Refresh: No): setting preferences outside an application's container requires user-preference-write or file-write-data sandbox access
2020-05-13 00:16:34.973566+0300 ObjCTrial2[22685:368459] [User Defaults] Couldn't write values for keys (
    ApplicationAccessibilityEnabled
) in CFPrefsPlistSource<0x600002c04b80> (Domain: com.apple.Accessibility, User: kCFPreferencesCurrentUser, ByHost: No, Container: (null), Contents Need Refresh: No): setting preferences outside an application's container requires user-preference-write or file-write-data sandbox access
2020-05-13 00:16:35.112237+0300 ObjCTrial2[22685:368459] Metal API Validation Enabled
2020-05-13 00:16:35.288575+0300 ObjCTrial2[22685:368459] [plugin] AddInstanceForFactory: No factory registered for id <CFUUID 0x60000025a260> F8BB1C28-BAE8-11D6-9C31-00039315CD46
2020-05-13 00:16:35.332341+0300 ObjCTrial2[22685:368459]  HALC_ShellDriverPlugIn::Open: Can't get a pointer to the Open routine
2020-05-13 00:16:35.332764+0300 ObjCTrial2[22685:368459]  HALC_ShellDriverPlugIn::Open: Can't get a pointer to the Open routine
2020-05-13 00:16:35.333207+0300 ObjCTrial2[22685:368459]  HALC_ShellDriverPlugIn::Open: Can't get a pointer to the Open routine
2020-05-13 00:16:35.388869+0300 ObjCTrial2[22685:368459] [settings] AKSettings.swift:setSession(category:with:):309:Cannot set AVAudioSession.setActive to true Error Domain=AVAudioSessionErrorDomain Code=561015905 "Error: Cannot activate session when app is in background." UserInfo={NSLocalizedDescription=Error: Cannot activate session when app is in background.} (AKSettings.swift:setSession(category:with:):309)
2020-05-13 00:16:35.389063+0300 ObjCTrial2[22685:368459] [general] AKManager.swift:output:77:Could not set session category: Error Domain=AVAudioSessionErrorDomain Code=561015905 "Error: Cannot activate session when app is in background." UserInfo={NSLocalizedDescription=Error: Cannot activate session when app is in background.} (AKManager.swift:output:77)
2020-05-13 00:16:35.391460+0300 ObjCTrial2[22685:368459] [settings] AKSettings.swift:setSession(category:with:):309:Cannot set AVAudioSession.setActive to true Error Domain=AVAudioSessionErrorDomain Code=561015905 "Error: Cannot activate session when app is in background." UserInfo={NSLocalizedDescription=Error: Cannot activate session when app is in background.} (AKSettings.swift:setSession(category:with:):309)
2020-05-13 00:16:36.789774+0300 ObjCTrial2[22685:368459] Stop oscillators
2020-05-13 00:16:38.018679+0300 ObjCTrial2[22685:368459] Start oscillators
2020-05-13 00:16:38.488316+0300 ObjCTrial2[22685:368459] Stop oscillators
```

The code I'm using is from the `HelloObjectiveC` example app:

https://gist.github.com/0xmtn/480af10146e3ed00c7e203b288a980cf

If I use the commented code in the gist above, which I was experimenting with, the log becomes:

```
2020-05-13 00:08:13.104920+0300 ObjCTrial2[22643:364307] [User Defaults] Couldn't write values for keys (
    ApplicationAccessibilityEnabled
) in CFPrefsPlistSource<0x600002c00200> (Domain: com.apple.Accessibility, User: kCFPreferencesCurrentUser, ByHost: No, Container: (null), Contents Need Refresh: Yes): setting preferences outside an application's container requires user-preference-write or file-write-data sandbox access
2020-05-13 00:08:13.107042+0300 ObjCTrial2[22643:364307] [User Defaults] Couldn't write values for keys (
    FullKeyboardAccessFocusRingEnabled
) in CFPrefsPlistSource<0x600002c00200> (Domain: com.apple.Accessibility, User: kCFPreferencesCurrentUser, ByHost: No, Container: (null), Contents Need Refresh: No): setting preferences outside an application's container requires user-preference-write or file-write-data sandbox access
2020-05-13 00:08:13.109929+0300 ObjCTrial2[22643:364307] [User Defaults] Couldn't write values for keys (
    ApplicationAccessibilityEnabled
) in CFPrefsPlistSource<0x600002c00200> (Domain: com.apple.Accessibility, User: kCFPreferencesCurrentUser, ByHost: No, Container: (null), Contents Need Refresh: No): setting preferences outside an application's container requires user-preference-write or file-write-data sandbox access
2020-05-13 00:08:13.300604+0300 ObjCTrial2[22643:364307] Metal API Validation Enabled
2020-05-13 00:08:13.501454+0300 ObjCTrial2[22643:364307] [plugin] AddInstanceForFactory: No factory registered for id <CFUUID 0x60000025caa0> F8BB1C28-BAE8-11D6-9C31-00039315CD46
2020-05-13 00:08:13.559324+0300 ObjCTrial2[22643:364307]  HALC_ShellDriverPlugIn::Open: Can't get a pointer to the Open routine
2020-05-13 00:08:13.560049+0300 ObjCTrial2[22643:364307]  HALC_ShellDriverPlugIn::Open: Can't get a pointer to the Open routine
2020-05-13 00:08:13.560596+0300 ObjCTrial2[22643:364307]  HALC_ShellDriverPlugIn::Open: Can't get a pointer to the Open routine
2020-05-13 00:08:13.729940+0300 ObjCTrial2[22643:364307] [avas] AVAudioSession_MacOS.mm:258:-[AVAudioSession getChannelsFromAU:PortName:PortID:]: ERROR in getting channel layout for auScope 1768845428 element 1
2020-05-13 00:08:13.732381+0300 ObjCTrial2[22643:364307] [avas] AVAudioSession_MacOS.mm:258:-[AVAudioSession getChannelsFromAU:PortName:PortID:]: ERROR in getting channel layout for auScope 1768845428 element 1
2020-05-13 00:08:13.734886+0300 ObjCTrial2[22643:364307] [avas] AVAudioSession_MacOS.mm:258:-[AVAudioSession getChannelsFromAU:PortName:PortID:]: ERROR in getting channel layout for auScope 1869968496 element 0
2020-05-13 00:08:13.736923+0300 ObjCTrial2[22643:364307] Input Devices: (
    "<Device: CADefaultDeviceAggregate-22643-0 (CADefaultDeviceAggregate-22643-0)>"
)
2020-05-13 00:08:13.737038+0300 ObjCTrial2[22643:364307] Output Devices: (
    "<Device: CADefaultDeviceAggregate-22643-0 (CADefaultDeviceAggregate-22643-0)>"
)
2020-05-13 00:08:13.800938+0300 ObjCTrial2[22643:364307] [avas] AVAudioSession_MacOS.mm:258:-[AVAudioSession getChannelsFromAU:PortName:PortID:]: ERROR in getting channel layout for auScope 1768845428 element 1
2020-05-13 00:08:13.805909+0300 ObjCTrial2[22643:364307] [avas] AVAudioSession_MacOS.mm:258:-[AVAudioSession getChannelsFromAU:PortName:PortID:]: ERROR in getting channel layout for auScope 1869968496 element 0
2020-05-13 00:08:13.806204+0300 ObjCTrial2[22643:364307] Output Device: <Device: CADefaultDeviceAggregate-22643-0 (CADefaultDeviceAggregate-22643-0)>
2020-05-13 00:08:13.809111+0300 ObjCTrial2[22643:364307] [settings] AKSettings.swift:setSession(category:with:):309:Cannot set AVAudioSession.setActive to true Error Domain=AVAudioSessionErrorDomain Code=561015905 "Error: Cannot activate session when app is in background." UserInfo={NSLocalizedDescription=Error: Cannot activate session when app is in background.} (AKSettings.swift:setSession(category:with:):309)
2020-05-13 00:08:13.809341+0300 ObjCTrial2[22643:364307] [general] AKManager.swift:output:77:Could not set session category: Error Domain=AVAudioSessionErrorDomain Code=561015905 "Error: Cannot activate session when app is in background." UserInfo={NSLocalizedDescription=Error: Cannot activate session when app is in background.} (AKManager.swift:output:77)
2020-05-13 00:08:13.810804+0300 ObjCTrial2[22643:364307] [settings] AKSettings.swift:setSession(category:with:):309:Cannot set AVAudioSession.setActive to true Error Domain=AVAudioSessionErrorDomain Code=561015905 "Error: Cannot activate session when app is in background." UserInfo={NSLocalizedDescription=Error: Cannot activate session when app is in background.} (AKSettings.swift:setSession(category:with:):309)
2020-05-13 00:08:16.522940+0300 ObjCTrial2[22643:364307] Stop oscillators
2020-05-13 00:08:17.095790+0300 ObjCTrial2[22643:364307] Start oscillators
2020-05-13 00:08:17.547562+0300 ObjCTrial2[22643:364307] Stop oscillators
2020-05-13 00:08:17.940112+0300 ObjCTrial2[22643:364307] Start oscillators

```

My Audio Midi Setup is like this:

![image](https://user-images.githubusercontent.com/732147/81746145-a57eea80-94ae-11ea-9e9d-9d14f64de971.png)

I looked up for the error messages separately without any luck. I wonder if I'm missing something here or is it because of the prematurity of Catalyst?

**Comment by aure** [comment]
Did you try changing the format to 44.1KHz?

**Comment by 0xmtn** [comment]
Yes, the default was 44.1KHz, same errors. Then I changed it to 48.0KHz just to try, same thing again.

Does this have any significance? `Error: Cannot activate session when app is in background.` 

This is a maccatalyst app, runnning on "My Mac". Why the message says "app is in background"?

**Comment by aure** [comment]
Are you using SwiftUI? If so check my comments in the ContentView.swift file for drums:

https://github.com/AudioKit/AudioKit/blob/fddc10e2f30c20af56ddf2f771fc1bee16d5bfc6/Examples/iOS%2BCatalyst/Drums/Drums/ContentView.swift#L36

```
// Important to start AudioKit after the app has moved to the foreground on Catalyst
      self.conductor.start()
```

This may be an Apple bug, but you have to wait for things to appear before starting AudioKit, for now.
